Add Radius.Compute/containerImages resource type and Kubernetes Terraform recipe (#151)
radius-project/resource-types-contrib: Compute/containerImages/README.md

│ dynamic-rp Pod                                                   │
│                                                                  │
│  ┌──────────────────────┐         ┌────────────────────────┐    │
│  │ dynamic-rp container │ ──────► │ buildkitd sidecar      │    │
│  │ (runs Terraform      │  unix   │ (rootless, no privs)   │    │
│  │  recipe)             │ socket  │                        │    │
│  └──────────────────────┘         └───────────┬────────────┘    │
│                                                │                 │
└────────────────────────────────────────────────┼─────────────────┘
                                                 │ HTTPS push
                                                 ▼
                                    ┌────────────────────────┐
                                    │ user's container       │
                                    │ registry               │
                                    └────────────────────────┘
```

## Recipes

| Platform | IaC Language | Recipe Name | Stage |
|---|---|---|---|
| Kubernetes | Terraform | `recipes/kubernetes/terraform/main.tf` | Alpha |

## Recipe Parameters

> The platform engineer registers the recipe once per environment with a default registry path. Registry credentials are delivered separately as a Kubernetes Secret mounted into dynamic-rp by the Helm chart (see [Prerequisites](#prerequisites)); they are not recipe parameters.

Author: Fixed in c75cd7e. Trimmed the README to a thin overview (matching the Compute/containers/README.md pattern); the stale content that contradicted the schema/recipe is removed and the canonical developer docs now live solely in containerImages.yaml.

```bash
rad recipe register default \
  --resource-type Radius.Compute/containerImages \
  --template-kind terraform \
  --template-path "git::https://github.com/radius-project/resource-types-contrib.git//Compute/containerImages/recipes/kubernetes/terraform" \
  --parameters registry=ghcr.io/myorg
```

| Parameter | Required | Description |
|---|---|---|

> | `registry` | yes | Registry path images are pushed under, e.g. `ghcr.io/myorg`. The recipe composes `<registry>/<resource-name>:<tag>` to form the full image reference. May be overridden per-resource via `properties.registry`. |

Author: Fixed in c75cd7e. Trimmed the README to a thin overview (matching the Compute/containers/README.md pattern); the stale content that contradicted the schema/recipe is removed and the canonical developer docs now live solely in containerImages.yaml.

## Resource Properties

| Property | Required | Description |
|---|---|---|
| `environment` | yes | Radius Environment ID. |
| `application` | yes | Radius Application ID. |
| `tag` | no | Tag for the produced image. Defaults to a content-addressable digest (`sha256-<hash>`) computed from the build inputs. **Required** when `build.source` is a remote git URL. |

> | `registry` | no | Per-resource override of the recipe's `registry` parameter. |

Author: Fixed in c75cd7e. Trimmed the README to a thin overview (matching the Compute/containers/README.md pattern); the stale content that contradicted the schema/recipe is removed and the canonical developer docs now live solely in containerImages.yaml.

> | `build.source` | yes | Source location. Either a git URL of the form `git::https://…` (BuildKit clones inside the cluster) or, for local-development workflows, a path that the rad CLI uploads as a tarball. |

Author: Fixed in c75cd7e. Trimmed the README to a thin overview (matching the Compute/containers/README.md pattern); the stale content that contradicted the schema/recipe is removed and the canonical developer docs now live solely in containerImages.yaml.

| `build.dockerfile` | no | Path to the Dockerfile relative to the context. Defaults to `Dockerfile`. |

> | `build.platforms` | no | Target platforms (e.g. `["linux/amd64", "linux/arm64"]`). When omitted, the build targets the BuildKit sidecar's native architecture. Multi-platform builds use cross-compilation. |

Author: Fixed in c75cd7e. Trimmed the README to a thin overview (matching the Compute/containers/README.md pattern); the stale content that contradicted the schema/recipe is removed and the canonical developer docs now live solely in containerImages.yaml.

## Output Properties

| Output | Description |
|---|---|
| `properties.imageReference` | The full resolved image reference, e.g. `ghcr.io/myorg/myimage:sha256-d4f2…`. Reference this from `Radius.Compute/containers` resources so they pick up new digests automatically. |

## Prerequisites

1. **dynamic-rp installed with the BuildKit sidecar enabled.** The default Helm install enables it. On Kubernetes < 1.30 (or any cluster without `UserNamespacesSupport`), install with `--set dynamicrp.buildkit.psaMode=baseline`.

2. **Registry credentials provisioned as a Kubernetes Secret.** The platform engineer creates a Secret in the dynamic-rp namespace whose `config.json` key holds a Docker config-format credentials document, then points the chart at it via `--set dynamicrp.buildkit.credentialsSecret=<secret-name>`.

   ```bash
   kubectl create secret generic ghcr-credentials \
     --from-file=config.json=$HOME/.docker/config.json \
     --namespace radius-system
   helm upgrade radius radius/radius \
     --set dynamicrp.buildkit.credentialsSecret=ghcr-credentials

>    ```

Author: Fixed in c75cd7e. Trimmed the README to a thin overview (matching the Compute/containers/README.md pattern); the stale content that contradicted the schema/recipe is removed and the canonical developer docs now live solely in containerImages.yaml.

- Other Radius applications or external consumers may pull the same image reference.

Operators who need image cleanup should run a separate registry retention policy. Examples:

- **GHCR**: [`actions/delete-package-versions`](https://github.com/actions/delete-package-versions) on a schedule.
- **Azure Container Registry**: `az acr task` with `acr purge`.
- **Generic OCI**: `oras` plus a cron-style cleanup.

## Local testing

The contrib repo's generic `test-recipe.sh` harness cannot run this recipe in CI today: it does not pass per-recipe `--parameters`, and the kind cluster used for CI has no test registry. To test changes locally:

```bash
# 1. Stand up a kind cluster with the dynamic-rp + buildkit sidecar
rad install kubernetes --set dynamicrp.buildkit.enabled=true \
  --set dynamicrp.buildkit.psaMode=baseline \
  --set dynamicrp.buildkit.credentialsSecret=ghcr-credentials

# 2. Register this recipe pointing at your branch
rad recipe register default \
  --resource-type Radius.Compute/containerImages \
  --template-kind terraform \
  --template-path "git::https://github.com/<you>/resource-types-contrib.git//Compute/containerImages/recipes/kubernetes/terraform?ref=<branch>" \

>   --parameters registry=ghcr.io/<your-org>

Author: Fixed in c75cd7e. The Local Testing section (with the wrong commands and the nonexistent test/app.bicep) was removed when the README was trimmed.

# 3. Deploy the test fixture

> rad deploy Compute/containerImages/test/app.bicep

Author: Fixed in c75cd7e. Trimmed the README to a thin overview (matching the Compute/containers/README.md pattern); the stale content that contradicted the schema/recipe is removed and the canonical developer docs now live solely in containerImages.yaml.

## Overview

> The Radius.Compute/containerImages Resource Type builds a container image from source and pushes it to a container registry. The resulting image reference can be consumed by `Radius.Compute/containers` resources within the same Application.

Author: ```suggestion
The Radius.Compute/containerImages Resource Type builds a container image from source and pushes it to a container registry.
```

Author: Applied in ceed4f4.

> ## Overview
> The Radius.Compute/containerImages Resource Type builds a container image from source and pushes it to a container registry.
> 
> Developer documentation is embedded in the Resource Type definition YAML file. Developer documentation is accessible via `rad resource-type show Radius.Compute/containerImages`.

Reviewer: ```suggestion
## Overview
The Radius.Compute/containerImages Resource Type builds a container image from source and pushes it to a container registry.

Builds run on the Radius control plane inside the dynamic-rp Pod using a rootless BuildKit sidecar. There is no host Docker socket, no privileged Pod, and no per-node host preparation. The Recipe use BuildKit by invoking the `buildctl` CLI mounted into the dynamic-rp container; the in-cluster buildkitd sidecar exposes its gRPC API on Pod loopback TCP.

Developer documentation is embedded in the Resource Type definition YAML file. Developer documentation is accessible via `rad resource-type show Radius.Compute/containerImages`.
```

Author: Applied in 34cc308 (with the typo fixed: 'The Recipe uses BuildKit...').

> ## Overview
> The Radius.Compute/containerImages Resource Type builds a container image from source and pushes it to a container registry.
> 
> Developer documentation is embedded in the Resource Type definition YAML file. Developer documentation is accessible via `rad resource-type show Radius.Compute/containerImages`.
> 
> ## Recipes
> 

Reviewer: ```suggestion

Prerequisites:

Using the containerImages resource requires platform engineers to configure the containerImages recipe with the target OCI registry. Developers cannot use containerImages without these steps complete.

1. The Radius Environment or Recipe Pack must define a Recipe parameter `registry` with the target OCI registry hostname.

2. If the registry requires authentication, a Radius secret resource must be created, then the `registrySecretName` Recipe parameter set on the Environment or Recipe Pack.

3. If using Kubernetes < 1.30, Radius must be installed with `--set dynamicrp.buildkit.psaMode=baseline`.

For example:

```bicep
extension radius

param registryUsername string
@secure()
param registryPassword string

resource env 'Radius.Core/environments@2025-08-01-preview' = {
  name: 'default'
  properties: {
    recipePacks: [ recipes.id ]
  }
}

resource recipes 'Radius.Core/recipePacks@2025-08-01-preview' = {
  name: 'container-images-recipe'
  properties: {
    recipes: {
      'Radius.Compute/containerImages': {
        recipeKind: 'terraform'
        recipeLocation: 'git::https://github.com/radius-project/resource-types-contrib.git//Compute/containerImages/recipes/kubernetes/terraform'
        parameters: {
          registry: 'ghcr.io/my-org'
          registrySecretName: 'ghcr-creds'
        }
      }
    }
  }
}

resource ghcrCreds 'Radius.Security/secrets@2025-08-01-preview' = {
  name: 'ghcr-creds'
  properties: {
    environment: env.id
    kind: 'generic'
    data: {
      username: { value: registryUsername }
      password: { value: registryPassword }
    }
  }
}
```

## Recipes

```

Author: Applied in 34cc308 — moved the Prerequisites section into README.md verbatim, including the full Bicep example with the recipePack and `Radius.Security/secrets`.

A list of available Recipes for this Resource Type, including links to the Bicep and Terraform templates:

|Platform| IaC Language| Recipe Name | Stage |
|---|---|---|---|
| Kubernetes | Terraform | recipes/kubernetes/terraform/main.tf | Alpha |

## Recipe Input Properties

Properties for the containerImages resource are provided to the Recipe via the [Recipe Context](https://docs.radapp.io/reference/context-schema/) object. These properties include:

> - `context.resource.properties.build.source` (string, required): The build context. Either a `git::https://...` URL or a local filesystem path. In the case of a local path, the rad CLI will package upload the source for the container build.

Author: Fixed in 34cc308. Removed the rad-CLI-upload reference from the README; the description now just says 'git URL or a local filesystem path' without naming an upload mechanism that doesn't yet exist.

> - `context.resource.properties.build.source` (string, required): The build context. Either a `git::https://...` URL or a local filesystem path. In the case of a local path, the rad CLI will package upload the source for the container build.

Author: Fixed in 34cc308. Removed the rad-CLI-upload reference from the README; the description now just says 'git URL or a local filesystem path' without naming an upload mechanism that doesn't yet exist.

> - `context.resource.properties.build.source` (string, required): The build context. Either a `git::https://...` URL or an absolute filesystem path available to the recipe runtime.

Author: ```suggestion
- `context.resource.properties.build.source` (string, required): The build context. Either a `git::https://...` URL or a local filesystem path. In the case of a local path, the rad CLI will package upload the source for the container build.
```

Author: Applied in ceed4f4.

- `context.resource.properties.build.dockerfile` (string, optional): Path to the Dockerfile relative to the build context. Defaults to `Dockerfile`.
- `context.resource.properties.build.platforms` (array of string, optional): Target platforms (e.g. `["linux/amd64", "linux/arm64"]`) for the multi-arch image. Defaults to `["linux/amd64", "linux/arm64"]`. Multi-arch builds require a cross-compile-friendly Dockerfile.
- `context.resource.properties.build.args` (object, optional): Map of `--build-arg` values passed to the build.

> - `context.resource.properties.tag` (string, optional): Explicit image tag. Required when `build.source` is a git URL. Defaults to a content-addressable tag (`sha256-<hash>`) for filesystem contexts.

Author: why is it required when it's a git URL? why wouldn't it just use a content-addressable tag in this case too

Author: Good question — fixed in ceed4f4. The original rationale was that we couldn't compute a hash from a remote tree, but we don't need to: the resolved git URL already contains the ref and subdir (from the BuildKit fragment), so hashing `{url, dockerfile, platforms, args}` gives a deterministic tag that changes only when one of those inputs changes. Same content-addressable behavior as local contexts; the `validate_git_tag` precondition is gone and an explicit `properties.tag` is now optional everywhere.

## Overview
The Radius.Compute/containerImages Resource Type builds a container image from source and pushes it to a container registry.

Builds run on the Radius control plane inside the dynamic-rp Pod using a rootless BuildKit sidecar. There is no host Docker socket, no privileged Pod, and no per-node host preparation. The Recipe uses BuildKit by invoking the `buildctl` CLI mounted into the dynamic-rp container; the in-cluster buildkitd sidecar exposes its gRPC API on Pod loopback TCP.

> Developer documentation is embedded in the Resource Type definition YAML file. Developer documentation is accessible via `rad resource-type show Radius.Compute/containerImages`.

Author: Updated the PR description to match what's actually in this PR (no `test/app.bicep`). The test fixture moved to a follow-up so we can land it together with the CI harness enhancement that injects per-recipe `--parameters` (radius-project/resource-types-contrib#154) — without that, a fixture wouldn't be exercisable end-to-end here anyway.

## Prerequisites

Using the containerImages resource requires platform engineers to configure the containerImages Recipe with the target OCI registry. Developers cannot use containerImages without these steps complete.

1. The Radius Environment or Recipe Pack must define a Recipe parameter `registry` with the target OCI registry hostname.

> 

Author: Good call — fixed in 1a9fe63. Prerequisite #1 now reads "target registry prefix images are pushed under" and explicitly says it's a hostname optionally followed by a path (`ghcr.io` or `ghcr.io/my-org`), noting that the recipe appends `/<resource-name>:<tag>`. The Recipe Input Properties section already worded it correctly, so the README is now consistent.

2. If the registry requires authentication, a Radius secret resource must be created, then the `registrySecretName` Recipe parameter set on the Environment or Recipe Pack.

3. If using Kubernetes < 1.30, Radius must be installed with `--set dynamicrp.buildkit.psaMode=baseline`.

For example:

```bicep
extension radius

param registryUsername string
@secure()
param registryPassword string

resource env 'Radius.Core/environments@2025-08-01-preview' = {
  name: 'default'
  properties: {
    recipePacks: [ recipes.id ]
  }
}

resource recipes 'Radius.Core/recipePacks@2025-08-01-preview' = {
  name: 'container-images-recipe'
  properties: {
    recipes: {
      'Radius.Compute/containerImages': {
        recipeKind: 'terraform'
        recipeLocation: 'git::https://github.com/radius-project/resource-types-contrib.git//Compute/containerImages/recipes/kubernetes/terraform'
        parameters: {
          registry: 'ghcr.io/my-org'
          registrySecretName: 'ghcr-creds'
        }
      }
    }
  }
}

resource ghcrCreds 'Radius.Security/secrets@2025-08-01-preview' = {
  name: 'ghcr-creds'
  properties: {
    environment: env.id

>     kind: 'generic'

Reviewer: ```suggestion
```

Author: Already removed in 3951bf7 — resolving.
